### PR TITLE
[JUJU-2615] Clean up entity parsing for debug-log

### PR DESCRIPTION
**Pull request by jack-w-shaw**
Clean up entity parsing for debug-log

Show warnings when filters filter non log giving sources

Also remove unsupported feature where debug-log filtering allows
filtering by regex, but only sometimes. Now only raw strings or
wildcards are supported

Do this by escaping regex then unescaping the wildcard when querying
Mongo. Hopefully changes to the client _should_ mean this is superfulous.

NOTE: queries can include `*` wildcard, so are not always valid tags or resource
names

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

Bootstrap a machine controller, deploy some apps and try
- `juju debug-log -i ubuntu` should show logs for all ubuntu units
- `juju debug-log -i "#p[kl;"` should show a warning
- `juju debug-log -i "*"` should show all logs
- `juju debug-log -i machine-*` should show all machine logs
- `juju debug-log -i "\w+-ubuntu"` should return a warning

Bootstrap a k8s controller and try, deploy some apps and try:
- `juju debug-log -i ubuntu` should show logs for the ubuntu application
- `juju debug-log -i 0` should return a warning
- `juju debug-log -i machine-0` should return a warning
- `juju debug-log -i ubuntu/0` should return a warning

## Bug reference

https://bugs.launchpad.net/juju/+bug/1977670

### CODE REVIEW

File: cmd/juju/commands/debuglog.go [hunk 1]
```diff
@@ -213,36 +213,49 @@ func (c *debugLogCommand) processEntities(isCAAS bool, entities []string) []stri
...
+	results := make([]string, len(entities))
 	for i, entity := range entities {
-		// A stringified unit or machine tag never match their "IsValid"
+		// Filter out valid tags early
+		if tag, err := names.ParseTag(entity); err == nil {
+			if tag.Kind() == names.UnitTagKind || tag.Kind() == names.MachineTagKind || tag.Kind() == names.ApplicationTagKind {
+				results[i] = entity
+				continue
+			}
+			logger.Warningf("Tag %s of kind %s is not a supported filter for debug-log", entity, tag.Kind())
```
Comment (hmlanigan) [inline cmd/juju/commands/debuglog.go:224]: We need to reword this to make it more user friendly. Most users have no idea what the juju tags are. 

Something like?
"WARNING charm-juju-qa-test is not a valid type for use with the --include flag. Expected examples include 0, machine-0, ubuntu, application-ubuntu, ubuntu/0"

Same for the other logger messages

File: cmd/juju/commands/debuglog.go [hunk 2]
```diff
@@ -213,36 +213,49 @@ func (c *debugLogCommand) processEntities(isCAAS bool, entities []string) []stri
...
+		}
+		// A stringified unit or machine tag are not themselves valid tags
 		// function from names, so if the string value passed in is a valid
 		// machine or unit, then convert here.
 		if names.IsValidMachine(entity) {
 			entity = names.NewMachineTag(entity).String()
 		} else if names.IsValidUnit(entity) {
 			entity = names.NewUnitTag(entity).String()
```
Comment (hmlanigan) [inline cmd/juju/commands/debuglog.go:232]: I believe these lines can be removed. names.ParseTag calls IsValidMachine and IsValidUnit while parsing the tag. I'm wondering if this entire section can be simplified.

Not sure how `names.UnitTagKind + "-" + entity + "-*"` is handled.

File: cmd/juju/commands/debuglog.go [hunk 3]
```diff
@@ -213,36 +213,49 @@ func (c *debugLogCommand) processEntities(isCAAS bool, entities []string) []stri
 	if entities == nil {
```
Comment (hmlanigan) [inline cmd/juju/commands/debuglog.go:213]: Is there a unit test for processEntities which can be updated?

**Review by hmlanigan — COMMENTED** [review]
This is an interesting one, many permutation to handle.

File: cmd/juju/commands/debuglog.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jack-w-shaw) [reply cmd/juju/commands/debuglog.go:224]: It is a little awkward. I decided against exiting out for 'safety' without really thinking through if it would actually keep anything safe. I'll exit out with an error I think

That will save a case of the weird mixed use of `ctx` and `logger` for logging in the client as well

File: cmd/juju/commands/debuglog.go (hunk 2 above)
Comment (jack-w-shaw) [reply cmd/juju/commands/debuglog.go:232]: I don't think so unfortunately, before `ParseTag` calls `IsValidX` it processes the tag (splits on the first `-`, etc.

We will only get to this step if a complete tag has not been provided such as `0`, `ubuntu`, `ubuntu/0` which return error on ParseTag. However, `0` will return true for `names.IsValidMachine`, which then needs to be parsed into a tag string

File: cmd/juju/commands/debuglog.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jack-w-shaw) [reply cmd/juju/commands/debuglog.go]: However, sometimes there can be some ambiguity whether our query is a already a tag or application name.

For instance, you could deploy ubuntu and name your application `charm-ubuntu` for instance

Comment (jack-w-shaw) [reply cmd/juju/commands/debuglog.go]: Agreed. Have re-written

File: cmd/juju/commands/debuglog.go (hunk 3 above)
Comment (jack-w-shaw) [reply cmd/juju/commands/debuglog.go:220]: Unit tests have now been handled

**Comment by jack-w-shaw** [comment]
This is now probably the most elegant solution.

However, we have returned to the behaviour where if you filter for something invalid, you receive a warning, and then tail an empty stream. This is a little clunky UX-wise

**Review by hmlanigan — COMMENTED** [review]
Code looks good, nice rewrite of the method.  I'm delayed in QA due to `apt` issues.

**Review by hmlanigan — APPROVED** [review]
QA, looks like this section requires updates on error vs warning etc in the PR description. As well, ubuntu is not a k8s charm.

1. `juju debug-log -i charm-ubuntu` does not produce a warning nor error as indicated. Technically it's a valid application name I believe.

**Comment by jack-w-shaw** [comment]
Oh I thought I had updated the descriptions. Apparently not though

Are you sure ubuntu isn't a k8s charm? `juju deploy ubuntu` for a k8s controller deploys just fine
